tidy3d/components/autograd/derivative_utils.py

```diff
@@ -0,0 +1,124 @@
+# utilities for autograd derivative passing
+from __future__ import annotations
+
+import numpy as np
+import pydantic.v1 as pd
+import xarray as xr
+
+from ..base import Tidy3dBaseModel
+from ..data.data_array import ScalarFieldDataArray
+from ..types import Bound, tidycomplex
+from .types import PathType
+from .utils import get_static
+
+# class passed to gradient computations
+# defined here to make things more compact, extendable, and better documented
+
+# we do this because importing these creates circular imports
```

Comment:
I guess this is fine for now but I think we should figure out how to do this more robustly... either reorganizing the types so that this doesn't happen or possibly use [forward references](https://peps.python.org/pep-0484/#forward-references).
Or maybe the best way would actually to `from __future__ import annotations` 🤔

Reply:
The issue is that MonitorData such as FieldData can't be imported into eg geometry and medium, where these derivates are computed. 

One option might be to extract the derivative computation to somewhere in web/api/autograd instead of being methods of the classes. However, i do kind of like having them defined in the classes as it keeps things organized. 

I think annotations will not really help here but could be wrong. Also I'm wary about messing with forward refs unless it's necessary.

Reply:
No this should be the exact use case for from __future__ import annotations then 😄 Postponed evaluation of annotations, i.e. you don't need to import the type you're using as a type hint. And it's essentially a forward ref under the hood.

Reply:
Yea maybe. I need to read up on forward refs a bit more.  Seems like a code smell to me though?

Reply:
Practically speaking there was no issue with just passing the field data as generic dicts of data arrays.